### ISSUE TITLE
global objects

### ISSUE THREAD

**Issue by benseitz** [getkart/kart#15]
Kirby has three global objects: `$site`, `$pages` & `$page` - what do we want for Kart?

`$kart->cart()` seems a bit strange, as does `$kart` for general stuff and `$cart` for the shopping cart. Any ideas?

**Comment by plaidscreen** [comment]
Even though I added a placeholder $kart global in kart.php, I'm actually going to try to completely avoid having globals at all. It makes sense for Kirby, but for this plugin, I can't think of any particular advantage besides code clarity - and that can be accomplished with Classes and static functions.

If we avoid having a single $kart object that the developer does `$kart->cart()` from, and just let the developer instantiate his own Cart object, then in theory, a site could be running multiple Carts in a single session - it would be up to theme developer to decide how he wants to store the information. Of course, I think we should provide helper functions to make the process super simple. Something like
```
$my_cart = Cart::load_stored("stored_cart_a","session");
// do some stuff with $my_cart
$my_cart->store_to("session");
```
This reminds me that I need to write a code style guide. I've got a few conventions I'm adamant about, but mostly I just want to make everybody's code look consistent, so I'm open to draft changes.

**Comment by benseitz** [comment]
This would make things more flexible, but I really would like some kind of Kirby styled API with globals.

It's an easy concept even for non-developers and would make a cheatsheet much more understandable.

**Comment by distantnative** [comment]
I agree with @benseitz. One of the beauties or rather "why is it easy to start with Kirby" is that it's API is rather intuitive and you can just start with $site, $pages and $page.

I don't think we would need ```$cart```, cause I think – even if it looks a lil odd – ```$kart->cart()``` would work just as fine. Plus developers could still just do ```$cart = $kart->cart();```.

Wouldn't it be also possible to have multiple carts with this API? As long as there is only 1 cart ```$kart->cart()```returns that one, but if you wanna add another one, you could do e.g. ```$kart->cart("session")``` and Kart would create a new one if a "session" one doesn't exist yet. And from there on you'd always need to  specify which one you wanna get ```$kart->cart("session1")->add()``` or ```$kart->cart("session2")->add()```.

On that note, I think ```$kart```would be the only global variable that is needed. Cause from this one developers could easily create their own. Especially since page controllers (and prob also models) will be an essential thing in building Kart themes anyways. So when you already do ```$product = $page;```, just add ```$cart = $kart->cart()``` or ```$customer = $kart->customer(); ```.

I just feel that something like
```php
$my_cart = Cart::load_stored_from("session") || new Cart();
// do some stuff with $my_cart
$my_cart->store_to("session");
```
isn't really the Kirby-way of realizing it.

**Comment by plaidscreen** [comment]
I guess that makes sense, it's probably what everyone expects, and avoids having to use namespaces in the theme

**Comment by benseitz** [comment]
I totally agree with @distantnative!
Also `$kart->cart("wishlist")` sounds like a nice feature ;)

**Comment by benseitz** [comment]
And since you mentioned `$kart->customers()` do we need that or would it be possible to extend Kirby's `$site->users()´?

**Comment by distantnative** [comment]
I know that we could use Kirby's users as you can add fields and stuff, just no panel editing atm.
Still not sure or indifferent tho, if we should.

But I'd still provide some wrappers in Kart. ```$kart->customers()``` would then just filter ```$site->users()``` to return all users with the specific Kart customer role. But even more it would be nice to have ```$kart->customer()``` which tries to get cookie/session information if a customer is still logged in and would return the customer object then, otherwise false (meaning that either the customer has to log in, create a customer account or use checkout without registering an account)

**Comment by distantnative** [comment]
something like ```$kart->cart("wishlist")```sounds sweet and – in that way it seems like – would not even need to be a real feature that Kart needs to support, but just something that the theme developer can add by using 2 cart instances e.g. ```$kart->cart("wishlist")``` and ```$kart->cart("order")```

On the other hand...
How do we store carts? So far I thought they'd be based on a cookie/session and expire. So a cart instance would be always linked to a customer, logged in or without an account. But a wishlist cart would need to be persistent and not expire and only available for logged in customers.

**Comment by benseitz** [comment]
I think both the cart and a wish list could be based on cookies/sessions if the user is logged out - But as soon someone logs in, carts are stored with the user-object.
That way you could store carts and wish lists with your account information and access them, even if you log in from a different device. 

Using the built in user/role management of kirby would save us some time and trouble and provide more security?

**Comment by benseitz** [comment]
Plus: Analyzing what customers put in their carts but didn't buy is a great opportunity for store owners to improve their shop.

**Comment by plaidscreen** [comment]
I was AFK most of the day but keeping track of it via my phone - This was/is a great discussion guys. As I was reading and thinking about your comments, made me totally change my opinion about some of the ways I wanted to implement major features. We are going to have a much better product in the end than if I had tackled this project alone.

**Comment by benseitz** [comment]
I think we work really well together :) And with combined resources and brain power this will turn out much better, than if any of us did this on their own :) I'm really happy with our project so far!
